### PR TITLE
adds solr rake tasks and builds out SolrWrapper API to support use outside of automated testing

### PR DESCRIPTION
**Pull request by flyingzumwalt**
For reference, these additions and changes were motivated by needs for use in https://github.com/dtulibrary/toshokan/blob/suggesters/lib/tasks/solr.rake and related projects.

### CODE REVIEW

File: lib/solr_wrapper/instance.rb [hunk 1]
```diff
@@ -41,9 +45,10 @@ def wrap(&_block)
     ##
     # Start Solr and wait for it to become available
     def start
-      extract
+      raise_error_unless_extracted
```
Comment (cbeer) [inline lib/solr_wrapper/instance.rb]: Can you say something about this change? I feel like it's a feature that things were automatically extracted as-needed.

File: lib/solr_wrapper/instance.rb [hunk 2]
```diff
@@ -204,55 +240,51 @@ def validate!(file)
...
+    # @param [Array] boolean_flags to include in command (ie. "-c")
+    # @return [StringIO] an IO object for the executed shell command
     # @see https://github.com/apache/lucene-solr/blob/trunk/solr/bin/solr
-    def exec(cmd, options = {})
-      output = !!options.delete(:output)
-      args = [solr_binary, cmd] + solr_options.merge(options).map { |k, v| ["-#{k}", "#{v}"] }.flatten
-
+    # @example start solr in cloud mode on port 8983
+    #   exec('start', {p: '8983'}, ["-c"])
+    def exec(cmd, options = {}, boolean_flags = [])
```
Comment (cbeer) [inline lib/solr_wrapper/instance.rb]: I'm not thrilled about adding a third argument just for boolean values. Do you think we could just capture option values that are the boolean true and map them appropriately? We might call this with, e.g.:

```ruby
o.exec(port: 8983, c: true)
```

File: lib/solr_wrapper/instance.rb [hunk 3]
```diff
@@ -204,55 +240,51 @@ def validate!(file)
...
+      boolean_flags ||= []
+      args = [solr_binary, cmd] + boolean_flags + solr_options.merge(options).map { |k, v| ["-#{k}", "#{v}"] }.flatten
       if IO.respond_to? :popen4
         # JRuby
         env_str = env.map { |k, v| "#{Shellwords.escape(k)}=#{Shellwords.escape(v)}" }.join(" ")
         pid, input, output, error = IO.popen4(env_str + " " + args.join(" "))
         @pid = pid
         stringio = StringIO.new
-        if verbose? && !output
+        IO.copy_stream(output, stringio)
```
Comment (cbeer) [inline lib/solr_wrapper/instance.rb]: I'm curious why we're dropping the `output` option?

File: lib/solr_wrapper/instance.rb [hunk 4]
```diff
@@ -295,11 +327,15 @@ def download_path
...
 
     def default_download_path
-      File.join(Dir.tmpdir, File.basename(download_url))
+      File.join(download_dir, File.basename(download_url))
     end
 
-    def solr_dir
-      @solr_dir ||= options.fetch(:instance_dir, File.join(Dir.tmpdir, File.basename(download_url, ".zip")))
+    def download_dir
+      if @download_dir.nil?
```
Comment (cbeer) [inline lib/solr_wrapper/instance.rb]: Could we use the memoization pattern here?

```ruby
@download_dir ||= ...
```

File: lib/solr_wrapper/instance.rb (hunk 1 above)
Comment (flyingzumwalt) [inline lib/solr_wrapper/instance.rb]: I don't remember why I removed that.  I think it had to do with annoyances that I ended up fixing elsewhere.

File: lib/solr_wrapper/instance.rb (hunk 2 above)
Comment (flyingzumwalt) [inline lib/solr_wrapper/instance.rb]: done

File: lib/solr_wrapper/instance.rb (hunk 3 above)
Comment (flyingzumwalt) [inline lib/solr_wrapper/instance.rb]: It wasn't actually doing anything meaningful other than unnecessarily preventing you from having access to the output in some cases.  Also, you need to at least capture the output within the method call in order return meaningful error messages.  Once you have the output in hand, there's not much point in _not_ returning it.

File: lib/solr_wrapper/instance.rb (hunk 4 above)
Comment (flyingzumwalt) [inline lib/solr_wrapper/instance.rb]: done.

**Comment by cbeer** [comment]
Thanks. In #9, I took these commits and tried to get the jruby build working again. I had to back out the output/verbose changes, which was causing me some trouble.

I also changed a couple things about the API:

- using `SolrWrapper.defaut_instance_options` to set default options (instead of the SOLR_OPTIONS constant)
- renaming `validate_existing` to just `validate`, and leaving it enabled by default.

**Pull request closed without merging**